rts/System/Sound/OpenAL/MusicStream.cpp

```diff
@@ -0,0 +1,316 @@
+//   passing that along to the callbacks via COggStream{::Play}
+// CFileHandler fileBuffers[NUM_AUDIO_CHANNELS];
+static CFileHandler fileBuffer("", "");
+
+// FIXME these parts of MusicStream are outside the class because every instance
+// of CSoundSource has a copy of MusicStream class but only one is ever active
+static constexpr unsigned int BUFFER_SIZE = 512 * 1024; // 512KB
+static std::variant<OggDecoder, Mp3Decoder> decoder;
+
+
+MusicStream::MusicStream(ALuint _source)
+	: pcmDecodeBuffer(nullptr)
+	, source(_source)
+	, format(AL_FORMAT_MONO16)
+	, stopped(true)
+	, paused(false)
+{
+	std::fill(buffers.begin(), buffers.end(), 0);
+}
+
+MusicStream::~MusicStream()
+{
+	Stop();
+	spring::SafeDeleteArray(pcmDecodeBuffer);
+}
+
+
+MusicStream& MusicStream::operator=(MusicStream&& rhs) noexcept
+{
+	if (this != &rhs) {
+		spring::SafeDeleteArray(pcmDecodeBuffer);
+		pcmDecodeBuffer = rhs.pcmDecodeBuffer;
+		rhs.pcmDecodeBuffer = nullptr;
+
+		for (auto i = 0; i < buffers.size(); ++i) {
+			std::swap(buffers[i], rhs.buffers[i]);
+		}
+
+		std::swap(source, rhs.source);
+		std::swap(format, rhs.format);
+
+		std::swap(stopped, rhs.stopped);
+		std::swap(paused, rhs.paused);
+
+		std::swap(msecsPlayed, rhs.msecsPlayed);
+		std::swap(lastTick, rhs.lastTick);
+	}
+
+	return *this;
+}
+
+// open an Ogg stream from a given file and start playing it
+void MusicStream::Play(const std::string& path, float volume)
+{
+	// we're already playing another stream
+	if (!stopped)
+		return;
+
+	fileBuffer.Close();
+	fileBuffer.Open(path);
+
+	if (!fileBuffer.FileExists()) {
+		LOG_L(L_ERROR, "[MusicStream::Play] File doesn't exist: %s", path.c_str());
+		return;
+	}
+
+	if (fileBuffer.GetFileExt() == std::string_view{"mp3"}) {
+		decoder = Mp3Decoder();
+	} else {
+		decoder = OggDecoder();
+	}
+
+	const bool loaded = std::visit([&](auto&& d) {
+			return d.LoadData(fileBuffer.GetBuffer().data(), fileBuffer.FileSize());
+			} , decoder);
+	if (!loaded) {
+		LOG_L(L_ERROR, "[MusicStream::Play] Could not load file: %s", path.c_str());
+		source = 0; // invalidate
+		assert(!Valid());
+		return;
+	}
+
+	int channels = std::visit([&](auto&& d) { return d.GetChannels(); }, decoder);
+	format = (channels == 1) ? AL_FORMAT_MONO16 : AL_FORMAT_STEREO16;
+
+	totalTime = std::visit([&](auto&& d) { return d.GetTotalTime(); }, decoder);
+
+	alGenBuffers(2, buffers.data());
+	CheckError("[MusicStream::Play][1]");
+
+	if (!StartPlaying()) {
+		ReleaseBuffers();
+	} else {
+		stopped = false;
+		paused  = false;
+	}
+
+	CheckError("[MusicStream::Play][2]");
+}
+
+// stops the currently playing stream
+void MusicStream::Stop()
+{
+	if (stopped)
+		return;
+
+	ReleaseBuffers();
+
+	msecsPlayed = spring_nulltime;
+	lastTick = spring_gettime();
+
+	source = 0;
+	format = 0;
+	totalTime = 0.0f;
+
+	assert(!Valid());
+}
+
+// clean up the OpenAL resources
+void MusicStream::ReleaseBuffers()
+{
+	stopped = true;
+	paused = false;
+
+#if 0
+	EmptyBuffers();
```

Comment:
perhaps remove? the comment in the other branch doesn't seem to say this is going to be useful

Reply:
I'm hesitant to change anything there just yet. I encountered sound cracking from music component when the music player widget enables "turn on music on loading" (starts plays music once the widget is loaded which happens mid loading screen) so the implementation there does not seem to be perfect. The sound cracking was already occuring on previous spring releases.